carbon-design-system/carbon
# test(components): add e2e tests for selectors to detect changes

Closes #8520 

This PR adds in a simple snapshot-based strategy for detecting changes to the selectors that we ship as part of our Sass/CSS files. It uses our compiled CSS file (`carbon-components/css/carbon-components.css`), parses it into an AST, and then finds all the CSS Rules that exist in the stylesheet.

We grab the selectors from each rule and then snapshot them in a test. When selectors change, this snapshot will fail and the release team will be notified. Changes that are intended will then be added to our changelogs so teams know what selectors to be on the lookout for since they have been removed or replaced.

This is the initial implementation of this approach to see if this strategy will work. As we learn more about it, we can change the strategy to better suite the use cases that we have.

#### Changelog

**New**

- Add an e2e test for selectors from `carbon-components`

**Changed**

**Removed**

#### Testing / Reviewing

- Verify that the snapshot produces an array of all the selectors from our styles
- Edit an existing sass file by removing or changing the name of a selector
- Build `carbon-components`
- Run `yarn test:e2e`, it should report a difference in the snapshot for the selectors of `carbon-components`
